src/plugins/event_annotation/common/fetch_event_annotations/request_event_annotations.ts

```diff
@@ -0,0 +1,357 @@
+        },
+      },
+    }));
+
+    const aggregations = [
+      annotationsFilters,
+      dateHistogram,
+      count,
+      timefieldTopMetric,
+      ...fieldsTopMetric,
+    ];
+
+    const aggConfigs = aggs.createAggConfigs(dataView, aggregations?.map((agg) => agg.value) ?? []);
+    return {
+      esaggsParams: { dataView, aggConfigs, timeFields: [group.timeField] },
+      fieldsColIdMap:
+        group.allFields?.reduce<Record<string, string>>(
+          (acc, fieldName, i) => ({
+            ...acc,
+            [fieldName]: `col-${i + 4}-${i + 5}`,
+          }),
+          {}
+        ) || {},
+    };
+  });
+};
+
+function regroupForRequestOptimization(
+  { groups }: FetchEventAnnotationsArgs,
+  input: ExpressionValueSearchContext | null
+) {
+  const outputGroups = groups
+    .map((g) => {
+      return g.annotations.reduce<Record<string, ManualGroup | QueryGroup>>((acc, current) => {
+        if (current.isHidden) {
+          return acc;
+        }
+
+        if (isManualAnnotation(current)) {
+          if (!isInRange(current, input?.timeRange)) {
+            return acc;
+          }
+          const manualSubgroup = acc.manual as ManualGroup | undefined;
+          return {
+            ...acc,
+            ['manual']: manualSubgroup
+              ? {
+                  ...manualSubgroup,
+                  annotations: [...manualSubgroup.annotations, current],
+                }
+              : {
+                  type: 'manual',
+                  annotations: [current],
+                },
+          };
+        } else {
+          const key = `${g.dataView.value.id}-${current.timeField}`;
+          const subGroup = acc[key] as QueryGroup;
+          if (subGroup) {
+            let allFields = [...(subGroup.allFields || []), ...(current.extraFields || [])];
+            if (current.textField) {
+              allFields = [...allFields, current.textField];
+            }
+            return {
+              ...acc,
+              [key]: {
+                ...subGroup,
+                allFields: [...new Set(allFields)],
+                annotations: [...subGroup.annotations, current],
+              },
+            };
+          }
+
+          let allFields = current.extraFields || [];
+          if (current.textField) {
+            allFields = [...allFields, current.textField];
+          }
+
+          return {
+            ...acc,
+            [key]: {
+              type: 'query',
+              dataView: g.dataView,
+              timeField: current.timeField,
+              allFields,
+              annotations: [current],
+            },
+          };
+        }
+      }, {});
+    })
+    .reduce((acc, currentGroup) => {
+      Object.keys(currentGroup).forEach((key) => {
+        if (acc[key]) {
+          const currentSubGroup = currentGroup[key];
+          const accSubGroup = acc[key];
+
+          if (isManualSubGroup(currentSubGroup) || isManualSubGroup(accSubGroup)) {
```

Comment:
> Will there ever be a case where currentSubGroup is manual, but accSubGroup is not? Seems like accSubGroup will always match the type of currentSubGroup by virtue of the algorithm?

No but I couldn't make the typings right 😅 

> Also, will there ever be more than one manual subgroup? If not, you could consider (not sure) whether or not it is clearer to handle the manual subgroup before the loop instead of having this if case that will only ever be relevant to a single subgroup.

But that's not single subgroup - we get one single subgroup per every group that can contain manual annotations that later on we have to flatten... The alternative is to run all the looping twice and just filter in different places. 

I left the `subGroup` as a `manual` or `dataView-timefield` group of single group. I switched the `accSubgGroup` as `requestGroup`, that makes more sense 👌🏼